### PR TITLE
Move DSSE Implementation to securesystemslib

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
Fixes: #370

### Description of the changes being introduced by the pull request:
This pull requests aims to add DSSE Envelope to securesystemslib. Moves DSSE implementation from [in-toto/securesystemslib](https://github.com/in-toto/securesystemslib) to here.

DSSE Envelope currently have three APIs
- `pae()` -> to generate pre auth encoding.
- `sign()` -> sign DSSE pre auth encoding.
- `verify()` -> verify DSSE pre auth encoding with n Keys.
sign and verify methods of DSSE currently utilizes `Key` and `Singer` provided by securesystemslib.

Serialization module commits are remaining, will be added to this PR soon.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: securesystemslib/metadata.py
Comment (lukpueh) [inline securesystemslib/metadata.py:43]:
```suggestion
        """Creates a DSSE Envelope from its JSON/dict representation.
```

**Review by lukpueh — COMMENTED** [review]
This is super cool, @PradyumnaKrishna! I only gave it a high-level review just now and diffed it to downstream master, which I had already reviewed at some point. You did a very nice job fixing conflicts with all the upstream signer developments! 

Please add serialization parts and we can invite other maintainers to take a look. 

I think it is feasible to review all in one PR after all. The advantage of a single PR is that we can more easily test in-toto (https://github.com/in-toto/in-toto/pull/503) and python-tuf (I'd like to try using this there as well) against that PR, and update if needed.

**Comment by lukpueh** [comment]
Really cool, @PradyumnaKrishna! I think we can work with this. I'm trying to integrate it with python-tuf right now. I want to see if the idea works there as well, before we move forward with this.

Do you feel like updating https://github.com/in-toto/in-toto/pull/503 in the meanwhile?

**Comment by PradyumnaKrishna** [comment]
> Do you feel like updating [in-toto/in-toto#503](https://github.com/in-toto/in-toto/pull/503) in the meanwhile?

Just two minor issues, first GPGSigner and GPGKey support is not present in this PR. Second, the naming of `sign` and `verify` methods.

**Comment by lukpueh** [comment]
FYI: I drafted a PR similar to https://github.com/in-toto/in-toto/pull/503 in https://github.com/theupdateframework/python-tuf/pull/2246, which adds DSSE to python-tuf.

It works pretty well as is, I'm just a little bit worried about the added complexity. But I don't think we can do anything about it here in securesystemslib. Let's see what python-tuf maintainers say.

**Comment by lukpueh** [comment]
> Just two minor issues, first GPGSigner and GPGKey support is not present in this PR....

Maybe we can land https://github.com/in-toto/in-toto/pull/503 without DSSE/GPG support?

> [...] Second, the naming of `sign` and `verify` methods.

 Maybe we don't even have to change the names here in secureystemslib. We can also just alias the functions with the names we need in the in-toto envelope subclass. E.g. in python-tuf I created a [`sign`](https://github.com/theupdateframework/python-tuf/blob/2bc97c41ecb844a52cd516f143678b9ba40dd821/tuf/api/metadata.py#L188-L197) alias, which uses the same name but has an additional argument. And  for `verify` I use a tuf-specific [`verify_delegate`](https://github.com/theupdateframework/python-tuf/blob/2bc97c41ecb844a52cd516f143678b9ba40dd821/tuf/api/metadata.py#L199-L213).

**Comment by adityasaky** [comment]
> Maybe we can land https://github.com/in-toto/in-toto/pull/503 without DSSE/GPG support?

I'm okay with merging without GPG support when using DSSE and picking that up right after. I think there's loads in here and in 503 that can tie into some other efforts that don't require GPG right away.

**Comment by PradyumnaKrishna** [comment]
> Maybe we can land [in-toto/in-toto#503](https://github.com/in-toto/in-toto/pull/503) without DSSE/GPG support?

I am wondering how I will do that. Currently, we have to pass a `Signer` (`GPGSigner`) to `Metablock` or `Envelope` to create a signature. I will have to put more effort in removing `GPGSigner` and `GPGKey` from that PR. Maybe you have a better idea?

File: securesystemslib/serialization.py
Comment (lukpueh) [inline securesystemslib/serialization.py:16]: FYI: I just closed this issue. Let's just stick to abc for the time being and remove the comment.

**Comment by lukpueh** [comment]
I gave this some thought and came to the following conclusion:

Let's implement something like backwards-compatible *_Legacy*GPGKey and *_Legacy*GPGSigner in in-toto and use it for traditional metadata (as described in https://github.com/secure-systems-lab/securesystemslib/issues/471) AND, for the time being, just not support gpg signature in DSSE metadata in in-toto.

This allows us to move forward with https://github.com/in-toto/in-toto/pull/503, without:
- mixing old and new gpg signing API (i.e. `securesystemslib.gpg.functions` vs. `securesystemslib.signer`)
- adding old signatures to DSSE, which does not have backwards-compatibility constraints yet, OR mixing old and new (spec-compliant) gpg signature formats (see [#450](https://github.com/secure-systems-lab/securesystemslib/issues/450) and [#448](https://github.com/secure-systems-lab/securesystemslib/issues/448))
- blocking on upstream gpg work (i.e. [#488](https://github.com/secure-systems-lab/securesystemslib/pull/488))

File: tests/test_metadata.py
Comment (jku) [inline tests/test_metadata.py:107]: does it really raise UnsupportedAlgorithmError?

**Review by jku — COMMENTED** [review]
Sorry, it's taken a while since it's a hard concept to really properly understand (after being so familiar with the TUF metadata style)... but I've now left some specific comments in code and there are a few high level discussion items below as well.

### Is securesystemslib the right place?

For the record I think it probably is... but the question may be worth discussing: securesystemslib is already a bit of mishmash collection of code and adding yet another fairly large system in there does not make the "idea of securesystemslib" any clearer. Contributors are easier to find when the code base is understandable and has a purpose...

### Is making payload an object attribute correct?

Payload is obviously available at deserialize time so it's easy to just store it in an attribute... but I'm imagining some code that modifies the payload in multiple steps and I wonder if it's good that the code can't really use the Envelope to store the payload during those steps -- or if it does, it then ends up serializing and deserializing the data constantly.

The alternative I'm imagining is the payload potentially being stored in a deserialized format and only serialized when the serialization is needed.... However, the current design would likely work ok with the python-tuf repository module so I'm leaning towards _yes, it's fine to enforce that payload is stored as serialized. User is expected to keep track of the deserialized payload object until they are ready to put it back in the envelope_

### Does threshold-verify really belong in this layer?

I did leave some rather nitpicky review comments in the verify() method already (because this code is so critical)... but as a higher-level question: If envelope as a concept does not even know what a threshold is, why does it contain a threshold-verify method? I think it maybe should not: Thresholds are something both python-tuf and intoto care about but that does not automatically mean Envelope should handle thresholds.

### Naming

The unfortunate naming clashes with python-tuf are not making it easier to understand the concept. Maybe they can't be avoided  but we could try? Like starting with module names, how about not having a `securesystemslib.metadata` module? `dsse` maybe?

File: securesystemslib/util.py
Comment (lukpueh) [reply securesystemslib/util.py:464]: Agreed, especially since we already have base64 utility functions in the public (?) `formats` module. Let's start a `_utils.py` or `_internal/utils.py` just with the b64enc and b64dec from this PR.

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:61]: Yeah, I think we kept it a list for simplicity, and because that's what the DSSE spec describes. I'm fine with using a dict internally, if there's a clear usage advantage. Though I don't know if the same requirements apply as for TUF `Metadata`, since I expect `Envelope` objects to usually have a much shorter lifetime (also see https://github.com/theupdateframework/python-tuf/pull/2246#pullrequestreview-1253567186).

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:173]: One could argue for both. If the user has full control over threshold and keys, then passing `len(keys) < threshold` can be seen as usage error.  If we see those parameters as part of the data to be verified, then VerificationError indeed makes more sense. 

IMO the latter is located at a different layer, e.g. TUF's `verify_delegate`.

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:192]: Generally, I agree with you. But the DSSE spec defines it like this: https://github.com/secure-systems-lab/dsse/blob/469c6e9fa6c4b7252fb71101084561cfc4cd0fa5/protocol.md#multi-signature-verification

Comment (jku) [reply securesystemslib/metadata.py:192]: wow that's ... unfortunate, dare I say broken

Comment (jku) [reply securesystemslib/metadata.py:192]: resolving since it's the spec that doesn't make sense

**Comment by jku** [comment]
> > Does threshold-verify really belong in this layer?
> ...
> As mentioned inline, we just implemented the [DSSE spec](https://github.com/secure-systems-lab/dsse/blob/31569a75dbe2b956f27b510394f5ff82a9958ec2/protocol.md#multi-signature-verification) here.

oh, I completely missed that. If this was a spec review I would strongly argue this has no place in the envelope format spec. Since this is not spec review, I guess it has to be part of the implementation. 

Although I will point out that `verify()` does not implement the last steps of the specified verify process anyway  (probably because they also make no sense in the envelope): 
> * Reject if PAYLOAD_TYPE is not a supported type.
> * Parse SERIALIZED_BODY according to PAYLOAD_TYPE. Reject if the parsing fails.
so arguably the threshold verification could be left to higher level as well :grimacing:

**Comment by jku** [comment]
> This also results in a different usage pattern, where Envelope is really just a metadata wrapper to be discarded after verification, and created right before signature creation respectively.

Oh this would be useful in documentation: it makes sense but isn't obvious at least not from TUF metadata background.

Also, this will make custom fields in the Envelope quite painful... but I think that might be a fair price to pay. I don't see much use for (long lived) custom fields anyway

File: securesystemslib/metadata.py
Comment (jku) [inline securesystemslib/metadata.py:162]: I resolved an earlier comment related to this... but I've decided I'm not happy about it: we should not return "accepted keys" that is not actually a complete list but just some of the valid keys.

This return value is also not part of the protocol spec. 

To be clear: I don't object to the method returning a set of keys, but I do object to it returning a vaguely defined set of keys: returning None would be better than that.

File: securesystemslib/metadata.py
Comment (jku) [reply securesystemslib/metadata.py:61]: The ephemeral nature of Envelope objects that you referred to elsewhere (the idea that you're not really expected to keep them around) probably means this is not useful.

File: securesystemslib/metadata.py
Comment (jku) [reply securesystemslib/metadata.py:173]: Yeah, this really relates to my higher level complaint that I think _all_ of the decisions made in this method are not really Envelope decisions.

So I'll close this specific complaint, the discussion about "who actually does threshold verification" is the real question...

**Comment by jku** [comment]
filed https://github.com/secure-systems-lab/dsse/issues/55 for spec discussion

File: tests/test_metadata.py
Comment (PradyumnaKrishna) [reply tests/test_metadata.py:107]: `self.key_dicts` contains RSA, ECDSA, and ED25519 keys. Two of the key raises `FormatError`, and one of them raises `UnsupportedAlgorithmError` for invalid scheme.

Comment (jku) [reply tests/test_metadata.py:107]: right :joy:

**Comment by jku** [comment]
I had a look at https://github.com/theupdateframework/python-tuf/pull/2292 and that does make it clear what the purpose of the serialization mixins and abstractions are. I'm not sure if it's _worth_ all of the engineering complexity to share the serialization code between Metadata and Envelope but it looks correct to me and I don't have a problem with it.

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:61]: Any opinions about this, @PradyumnaKrishna?

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:101]: Ping

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:140]: Works for me. Let's leave it to the user to `envelope.signatures.clear(); envelope.sign()`, to get `append=False`

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:144]: Let's use `sign` and `verify` here and resolve name clashes in the custom in-toto Envelope subclass.

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:153]: Ping

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:158]: Ping

File: securesystemslib/metadata.py
Comment (lukpueh) [reply securesystemslib/metadata.py:162]: Ping

Comment (lukpueh) [reply securesystemslib/metadata.py:162]: Let's return None then for the time being so that we provide a spec compliant implementation. Also, AFAICS, we don't currently need the return value in in-toto.

File: securesystemslib/serialization.py
Comment (lukpueh) [reply securesystemslib/serialization.py:16]: Ping

Comment (lukpueh) [reply securesystemslib/serialization.py:16]: Ping

File: securesystemslib/util.py
Comment (lukpueh) [reply securesystemslib/util.py:464]: Ping

**Review by lukpueh — CHANGES_REQUESTED** [review]
Looks like we can finally get this merged. @jku, [green-lighted](https://github.com/secure-systems-lab/securesystemslib/pull/487#issuecomment-1419021014) the serialization-related code and what's left for DSSE is commented inline.

Regarding naming, I take back my suggestion to create a *metadata* subpackage. It's unlikely that we will create a common traditional metadata class for tuf and in-toto (#272), which such a package would have been a good fit for. Because tuf and in-toto already have working metadata implementations, which they actually want to move away from in favor of DSSE.

So I suggest the following renames for this PR:
- util.py -> _internal/utils.py 
- metadata.py -> dsse.py
- test_metadata.py -> test_dsse.py

@PradyumnaKrishna, do you care for finalizing the PR?

**Comment by PradyumnaKrishna** [comment]
> @PradyumnaKrishna, do you care for finalizing the PR?

I will work on in-toto/in-toto#503 first, and then I will update this PR according to the in-toto integration.

**Comment by PradyumnaKrishna** [comment]
@lukpueh, are we going to use `create_sig()` and `verify_sigs()` instead of `sign()` and `verify()` method of DSSE?

**Comment by lukpueh** [comment]
> @lukpueh, are we going to use create_sig and verify_sigs instead of sign and verify for signing and verification method names of DSSE?

I suggest we keep using `sign` and `verify` here. We need to subclass `Envelope` in in-toto anyway. We can just create `Metablock`-compatible aliases for `sign` and `verify` there. WDYT?

**Comment by lukpueh** [comment]
>  I'm not sure if it's _worth_ all of the engineering complexity to share the serialization code between Metadata and Envelope but it looks correct to me and I don't have a problem with it.

I fear @jku has a point here. Reasons to add the serialization infrastructure (base- and json-de/serializers plus common methods to read and write from and to files), was to re-use it for in-toto, python-tuf and a generic securesystemslib dsse envelope.

BUT given that we are:
 - unlikely to use it in python-tuf (details in  [tuf#2292](https://github.com/theupdateframework/python-tuf/pull/2292)), and
 - don't have a use case for a standalone securesystemslib Envelope, and
 - are generally trying to reduce the engineering footprint in this library,

... I suggest we don't include the de/serialization code with this PR after all, but figure out de/serialization for in-toto independently (https://github.com/in-toto/in-toto/pull/503). 

This means `Envelope` loses its `_default_deserializer`, `_default_serializer` and `get_payload` methods.  But it can keep `from_dict` and `to_dict`. This is consistent with other metadata-related classes in securesystemslib.signer, and may be helpful for any custom de/serialization solution.

File: securesystemslib/dsse.py
Comment (lukpueh) [inline securesystemslib/dsse.py:121]: Minor nit: Sphinx renders `accepted_keys` as return type, which is incorrect. Besides, the name is an implementation detail of this method, so no need to mention it in the docstring.

Comment (lukpueh) [reply securesystemslib/dsse.py:121]: Also, @jku made a good point about the return value being too vaguely defined, and not part of the dsse spec ( see https://github.com/secure-systems-lab/securesystemslib/pull/487#discussion_r1092886518)

I suggest we either return `None` -- I checked in-toto, we don't use it there -- or we add a tiny bit more detail in the documentation to explains that these are only the first ``threshold`` of ``keys`` that verified signatures.

File: securesystemslib/dsse.py
Comment (lukpueh) [inline securesystemslib/dsse.py:104]: Looking at how we use this function in in-toto --  `verify(keys=[key], threshold=1)`, and building in-toto specific threshold check verification around it --  @jku's argument in secure-systems-lab/dsse#55 seems more convincing now.

That said, we can work with what we have and I'd really like to move forward. Let's just add comment that makes it very clear that this API isn't final yet. E.g...


```suggestion
        """Verify the payload with the provided Keys.
        
        NOTE: This API is experimental and might change (see secure-systems-lab/dsse#55)
```

File: securesystemslib/dsse.py
Comment (lukpueh) [reply securesystemslib/dsse.py:121]: ping: https://github.com/secure-systems-lab/securesystemslib/pull/487/files#r1113994810

Comment (lukpueh) [reply securesystemslib/dsse.py:121]: Let's just add the docs here and move on. What do you think about...
```suggestion
            A dict of the threshold of unique public keys that verified a signature.
```
?

**Review by lukpueh — CHANGES_REQUESTED** [review]
Let's wrap this up, it looks like this works well enough for our in-toto use case (https://github.com/in-toto/in-toto/pull/503). 

Before we merge and release, would you mind accepting my inline comments?

**Comment by lukpueh** [comment]
Thanks for the update, @PradyumnaKrishna! CI failure is unrelated (see #520). Merging. 🎉

**Review by lukpueh — APPROVED** [review]

**Pull request merged**
